### ISSUE TITLE
flag.Parse in init()

### ISSUE THREAD

**Issue by rkjdid** [soul9/glog#1]
small shortcoming, the flag.Parse() call forbids any new flag to be created, so other custom flags created after package is imported are ignored. and of course flag.Set has no efect if !flag.Parsed()..

so yeah, I don't have a workaround idea

**Comment by rkjdid** [comment]
actually flag.Set seems to work, without preceding flag.Parse()

**Comment by soul9** [comment]
but then I can not check whether logtofiles has been set before setting logtostderr

**Comment by rkjdid** [comment]
yes

**Comment by rkjdid** [comment]
we'd need a notify mechanism flag.Parse(), I don't see such thing there

**Comment by soul9** [comment]
there is a "notify" mechanism polling every 1/10th of a second up to a second before calling flags.Parse. This might need to go into a goroutine? Not sure how init() blocks startup.

**Comment by soul9** [comment]
yeah this'll block package initialization anyways, so it's completely useless. Let's just say if you import this package you log to stderr and that's it. It really makes no sense otherwise.

**Comment by soul9** [comment]
no that doesn't work either since we need flag.Parsed before lfag.Set. suggestions?

**Comment by rkjdid** [comment]
I managed to make Set work without flag.Parsed(), did not work for you ?

**Comment by rkjdid** [comment]
it kind of defaults the flag to true, then the parsing overrides it. testing this on BoolVar is tricky though cause I think -bool 0 just doesn't work, I think it was changed at some point and now boolvar needs to default false and you can only enable it

**Comment by soul9** [comment]
I can default a bool flag to true. I think you might be confusing boolvar and bool. boolvar defaults to the value of the var you pass it.

**Comment by rkjdid** [comment]
you can default it to true, but try setting it to false afterwards with commandline. I didn't manage

**Comment by rkjdid** [comment]
btw already encountered this usecase previously, that's my 2nd round (#3 :D)

**Comment by soul9** [comment]
you *need* to use the = to set it, it seems. from https://golang.org/pkg/flag
<pre>
-flag x  // non-boolean flags only
</pre>

**Comment by rkjdid** [comment]
thx. the equal notation

**Comment by soul9** [comment]
so if I set the flag in a package's init, it looks like the init is run before flag.Parse() in the main package's init. which is perfect.

**Comment by rkjdid** [comment]
exactly, I didn't realize the -bool=f solved the whole issue meanwhile
init's imports beforehand makes sense too
perhaps no reopen ?

**Comment by soul9** [comment]
wat?

**Comment by soul9** [comment]
this doesn't ufkcign work

**Comment by rkjdid** [comment]
it does though

**Comment by rkjdid** [comment]
let me try again.

**Comment by soul9** [comment]
no, it doesn't

**Comment by rkjdid** [comment]
works for me xD

**Comment by rkjdid** [comment]
quote: problem is the set doesn't set the flag in this case.

didn't read that. but I confirm it's ok on my side, also -logtostderr=f disables errlog

**Comment by soul9** [comment]
try this https://gist.github.com/soul9/bacecc1f62654dd11e91

**Comment by rkjdid** [comment]
abort "import ./", this is shit

**Comment by soul9** [comment]
:D

**Comment by soul9** [comment]
same problem without ./

**Comment by soul9** [comment]
this https://github.com/soul9/go-experiments

**Comment by soul9** [comment]
ufkc programming i'm going farming

**Comment by rkjdid** [comment]
haha, yeah that.

**Comment by soul9** [comment]
i've got a wonderful solution that uses 100%cpu time and still doesn't work without a workaround.

**Comment by soul9** [comment]
check go-experiments new commit has the solution

**Comment by rkjdid** [comment]
yeah, 100%, that with a sleep would be even dirtier and yet less dirty, muchfun

**Comment by rkjdid** [comment]
i still want an explanation why it looks like it's working good with -logtostderr x_x

**Comment by soul9** [comment]
because I bet you are not creating the flag in package main, and luckily soul9/glog init is run after golang/glog init

**Comment by rkjdid** [comment]
meh. ok

**Comment by rkjdid** [comment]
although no. this works in my init's test

```
func init() {
	fmt.Println(flag.Parsed()) // false
	flag.Set("logtostderr", "true")
	flag.Parse()
	fmt.Println(flag.Parsed()) // true
}
```

**Comment by rkjdid** [comment]
no import soul9/glog

**Comment by soul9** [comment]
see newest commit in go-experiments

**Comment by soul9** [comment]
right, that should work fine.

**Comment by soul9** [comment]
import golang/glog though?

**Comment by soul9** [comment]
I don't think there is a simple way to do this in an external package, though, so we'll need the boilerplate in our mains, which is fine. the import is one line, boilerplate is one line, really. at most four if you want to make it optional with an extra logtofiles flag.

**Comment by rkjdid** [comment]
yeah, that was the conclusion I was about to draw before you hit me with itdoesntwork, nice manoeuvre we managed here

**Comment by soul9** [comment]
"much ado about nothing"
                                         -- Albert Einstein

**Comment by soul9** [comment]
deleting repo. thanks for all the fish.
-- Marilyn Monroe

**Comment by rkjdid** [comment]
yet, more knowledge, ah..
